### PR TITLE
add closure for updating shipping option once selected

### PR DESCRIPTION
**Pull request by ScottAntonacAP**
This PR aims to allow the updating of a single shipping option once it has been selected (for instance updating the tax amount)

## Summary of Changes

- Add a new param to `CheckoutV2Options`: `enableSingleShippingOptionUpdate`. This will need to be set to true to use this feature.
- Add functionality to example app
- Add a new model class: `ShippingOptionUpdate` that only takes `id` and `Money` amounts
- post message to checkout as closure to update shipping option
- send back empty payload if updated shipping option is `nil`

## Notes
This PR contains the following breaking changes
- public type alias `ShippingOptionsDidChangeClosure` changes to `ShippingOptionDidChangeClosure` for clarity
- public protocol `CheckoutV2Handler`'s method `shippingOptionDidChange` now takes a second parameter `completion` of type `ShippingOptionCompletion`

### CODE REVIEW

File: Example/Example/Purchase/CheckoutHandler.swift [hunk 1]
```diff
@@ -10,19 +10,22 @@ import Afterpay
...
 
   private var checkoutTokenResultCompletion: TokenResultCompletion?
   private var shippingOptionsCompletion: ShippingOptionsCompletion?
+  private var shippingOptionCompletion: ShippingOptionCompletion?
 
   init(
     didCommenceCheckout: @escaping () -> Void,
-    onShippingAddressDidChange: @escaping (ShippingAddress) -> Void
+    onShippingAddressDidChange: @escaping (ShippingAddress) -> Void,
+    onShippingOptionChange: @escaping (ShippingOption) -> Void
```
Comment (benfterpay) [inline Example/Example/Purchase/CheckoutHandler.swift:24]: Should this method name include the word "did"? Here and in other places below.

```suggestion
    onShippingOptionDidChange: @escaping (ShippingOption) -> Void
```

File: Example/Example/Purchase/CheckoutHandler.swift [hunk 2]
```diff
@@ -10,19 +10,22 @@ import Afterpay
...
   ) {
     didCommenceCheckoutClosure = didCommenceCheckout
     onShippingAddressDidChangeClosure = onShippingAddressDidChange
+    onShippingOptionClosure = onShippingOptionChange
```
Comment (benfterpay) [inline Example/Example/Purchase/CheckoutHandler.swift:28]:
```suggestion
    onShippingOptionDidChangeClosure = onShippingOptionDidChange
```

File: Example/Example/Purchase/CheckoutHandler.swift [hunk 3]
```diff
@@ -48,6 +51,14 @@ final class CheckoutHandler: CheckoutV2Handler {
     shippingOptionsCompletion = nil
   }
 
-  func shippingOptionDidChange(shippingOption: ShippingOption) {}
+  func shippingOptionDidChange(shippingOption: ShippingOption, completion: @escaping ShippingOptionCompletion) {
+    shippingOptionCompletion = completion
+    onShippingOptionClosure(shippingOption)
```
Comment (benfterpay) [inline Example/Example/Purchase/CheckoutHandler.swift:56]:
```suggestion
    onShippingOptionDidChangeClosure(shippingOption)
```

File: Example/Example/Purchase/PurchaseFlowController.swift [hunk 4]
```diff
@@ -37,7 +37,8 @@ final class PurchaseFlowController: UIViewController {
 
     checkoutHandler = CheckoutHandler(
       didCommenceCheckout: logicController.loadCheckoutToken,
-      onShippingAddressDidChange: logicController.selectAddress
+      onShippingAddressDidChange: logicController.selectAddress,
+      onShippingOptionChange: logicController.selectShipping
```
Comment (benfterpay) [inline Example/Example/Purchase/PurchaseFlowController.swift:41]:
```suggestion
      onShippingOptionDidChange: logicController.selectShipping
```

**Review by benfterpay — CHANGES_REQUESTED** [review]
Suggesting some tweaks to the naming conventions used in the example app. Also just a request, can you please double-check that the distinction between "ShippingOption" (singular) and "ShippingOptions" (plural) is appropriate in all occurrences?

File: Example/Example/Purchase/CheckoutHandler.swift [hunk 5]
```diff
@@ -10,19 +10,22 @@ import Afterpay
 import Foundation
 
 final class CheckoutHandler: CheckoutV2Handler {
-
   private let didCommenceCheckoutClosure: () -> Void
   private let onShippingAddressDidChangeClosure: (ShippingAddress) -> Void
+  private let onShippingOptionClosure: (ShippingOption) -> Void
```
Comment (benfterpay) [inline Example/Example/Purchase/CheckoutHandler.swift:15]:
```suggestion
  private let onShippingOptionDidChangeClosure: (ShippingOption) -> Void
```

**Comment by ScottAntonacAP** [comment]
> can you please double-check that the distinction between "ShippingOption" (singular) and "ShippingOptions" (plural) is appropriate in all occurrences?

Changed one occurrence of this: `ShippingOptionsDidChangeClosure` to `ShippingOptionDidChangeClosure`.

**Review by benfterpay — APPROVED** [review]
Looks good @ScottAntonacAP.

File: Example/Example/Purchase/PurchaseLogicController.swift [hunk 6]
```diff
@@ -145,21 +147,40 @@ final class PurchaseLogicController {
...
       ),
     ]
 
     let result: ShippingOptionsResult = .success(shippingOptions)
     commandHandler(.provideShippingOptionsResult(result))
   }
 
+  func selectShipping(shippingOption: ShippingOption) {
+    switch shippingOption.id {
+    case "standard":
```
Comment (nesevis) [inline Example/Example/Purchase/PurchaseLogicController.swift:169]: Suggestion: It might be more readable for a potential user browsing the Example app if, in this case, we opt for a simpler `if/else` clause rather than a `switch` with one case.

**Review by nesevis — APPROVED** [review]
Looks good. Aside from the specific comment I left above, there is another issue, but Github won't let me do it inline:

You may want to add a `type_body_length` exception for swiftlint here:

https://github.com/afterpay/sdk-ios/blob/aeaec194764b07cc845d3b53b8e952772f7b9b3d/Sources/Afterpay/Checkout/CheckoutV2ViewController.swift#L14-L21

I would suggest that we could break apart the entire class and move protocol conformances to extensions, but it's not done anywhere else in the codebase, so adding an exception seems the pragmatic thing to do.

File: Example/Example/Purchase/PurchaseLogicController.swift (hunk 6 above)
Comment (ScottAntonacAP) [reply Example/Example/Purchase/PurchaseLogicController.swift]: addressed in 7a1290a

**Comment by ScottAntonacAP** [comment]
> add a `type_body_length` exception for swiftlint here to address the warning

addressed in 41dcf6e
agree that it would be better to break apart the class, potentially in a separate PR. for now have disabled the rule

**Comment by adamjcampbell** [comment]
@nesevis 
> I would suggest that we could break apart the entire class and move protocol conformances to extensions, but it's not done anywhere else in the codebase, so adding an exception seems the pragmatic thing to do.

@ScottAntonacAP 
> agree that it would be better to break apart the class, potentially in a separate PR

👋 hello it's the ghost of development past. Obviously go forward and make what ever changes you see fit as the current maintainers. Original context / reasoning to not do this was that I hold the belief that the perceived cleanliness of breaking the class into extensions actually decreases your ability to understand what a class does at a glance. You must now traverse the whole file or any `ViewController+Extention.swift` files to understand the scope of it's responsibilities.

Tl;dr I have the personal preference to see a class and it's responsibilities in a single entity. If I were to find that the class has too many I'd sooner break out that responsibility (e.g. extracted delegate) rather than forming pseudo small classes / extracted responsibilities within extensions.

Anyway do with that what you will and feel free to dismiss it wholesale 😆
